external/js/cothority/src/byzcoin/proof.ts

```diff
@@ -121,15 +123,36 @@ export default class Proof extends Message<Proof> {
             return new Error("invalid root");
         }
 
-        let publics = this.links[0].newRoster.getServicePublics(SkipchainRPC.serviceName);
-        let prev = this.links[0].to;
-
-        if (!prev.equals(genesisID)) {
+        const links = this.links;
+        if (!links[0].to.equals(genesisID)) {
             return new Error("first link must come from the genesis block");
         }
 
-        const links = this.links;
-        for (let i = 1; i < links.length; i++) {
+        let latestKnown = 0;
+        const latestBlocks = Proof.knownBlocks.get(genesisID);
+        if (latestBlocks) {
+            links.map((l) => l).reverse().find((link, index) => {
+                if (latestBlocks.find((cache) => cache.equals(link.to))) {
+                    latestKnown = links.length - index - 1;
+                    return true;
+                }
+                return false;
+            });
+        }
+
+        if (latestKnown === links.length - 1) {
+            return null;
+        }
+
+        let publics: Point[];
+        links.slice(0, latestKnown + 1).forEach((link) => {
```

Comment:
This could go reverse and exit at the first defined `newRoster`.

Reply:
Dang - I'm not even sure this works in all cases - will have to look at it later with a fresh head...

Reply:
It was wrong - with multiple roster changes this would've failed...